core/services/pipeline/runner.go

```diff
@@ -433,30 +444,74 @@ func (r *runner) ExecuteAndInsertFinishedRun(ctx context.Context, spec Spec, var
 		return 0, finalResult, nil
 	}
 
-	if runID, err = r.orm.InsertFinishedRun(r.orm.DB(), run, trrs, saveSuccessfulTaskRuns); err != nil {
+	if runID, err = r.orm.InsertFinishedRun(postgres.UnwrapGormDB(r.orm.DB()), run, saveSuccessfulTaskRuns); err != nil {
 		return runID, finalResult, errors.Wrapf(err, "error inserting finished results for spec ID %v", spec.ID)
 	}
 	return runID, finalResult, nil
 
 }
 
-func (r *runner) Run(ctx context.Context, run *Run, l logger.Logger, saveSuccessfulTaskRuns bool) (incomplete bool, err error) {
+func (r *runner) Run(ctx context.Context, run *Run, l logger.Logger, saveSuccessfulTaskRuns bool, fn func(tx *gorm.DB) error) (incomplete bool, err error) {
+	pipeline, err := r.initializePipeline(run)
+	if err != nil {
+		return false, err
+	}
+
+	async := pipeline.HasAsync()
+
+	err = postgres.GormTransactionWithDefaultContext(r.orm.DB(), func(tx *gorm.DB) error {
+		// avoid an extra db write if there is no async tasks present or if this is a resumed run
+		if async && run.ID == 0 {
+			now := time.Now()
+			// initialize certain task params
+			for _, task := range pipeline.Tasks {
+				switch task.Type() {
+				case TaskTypeETHTx:
+					run.PipelineTaskRuns = append(run.PipelineTaskRuns, TaskRun{
+						ID:            task.Base().uuid,
+						PipelineRunID: run.ID,
+						Type:          task.Type(),
+						Index:         task.OutputIndex(),
+						DotID:         task.DotID(),
+						CreatedAt:     now,
+					})
+				default:
+				}
+			}
+			if err = r.orm.CreateRun(postgres.UnwrapGorm(tx), run); err != nil {
+				return err
+			}
+
+		}
+
+		if fn != nil {
```

Comment:
When would `fn` be nil?

Reply:
I guess never, since we default to a no-op, I just hate dealing with potential nil pointer bugs. I'll remove the check?